File: C++/11_Strings/EASY_LONGEST_PALINDROME.cpp
/* Given a string which consists of lowercase or uppercase letters, find the length of the longest palindromes that can be built with those letters.

This is case sensitive, for example "Aa" is not considered a palindrome here.

Note:
Assume the length of given string will not exceed 1,010. */

// My solution
int longestPalindrome(string s) {
    map<char, int> counts;
    for (char c:s)
    {   if (counts.find(c) == counts.end())
            counts[c]=1;
        else counts[c]++;
    }
    
    int l = 0;
    for (auto it:counts)
    {   if (it.second % 2 == 0)
            l+=it.second;
        else l += it.second-1;
    }    
    
    if (l < s.size()) l++;
    return l;
}

File: Python/07_Hash_Table/easy_longest_palindrome.py
'''
Given a string which consists of lowercase or uppercase letters,
find the length of the longest palindromes that can be built with those letters.

This is case sensitive, for example "Aa" is not considered a palindrome here.
'''

'''
Time -> O(n)
Space -> O(n) [in dict]
'''
class Solution:
    def longestPalindrome(self, strng):

        # Create a frequency dictionary
        ref = {}
        for s in strng:
            if s in ref:
                ref[s] += 1
            else:
                ref[s] = 1

        ans = 0
        extra_char = False
        for k, v in ref.iteritems():
            # Use all characters if it's even
            if v % 2 == 0:
                ans += v
            else:
                ans += v-1
                extra_char = True

        # Instead of the following condition you can also say
        # if ans < len(strng)
        if extra_char:
            return ans + 1
        return ans
